### ISSUE TITLE
zip64 bug in LogicalZipFile

### ISSUE THREAD

**Issue by cushon** [classgraph/classgraph#841]
The zip implementation in `LogicalZipFile` contains a zip64 bug around handling the end of central directory record.

If an archive has a zip64 end of central directory record, and the value of a field in the non-zip64 end of central directory is `0xFFFF` or `0xFFFFFFFF`, the value is in the corresponding zip64 end of central directory field.

For example for the end of central directory size, from [APPNOTE.TXT](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) 4.4.23

> The size (in bytes) of the entire central directory. If an archive is in ZIP64 format and the value in this field is 0xFFFFFFFF, the size will be in the corresponding 8 byte zip64 end of central directory field.

The ClassGraph implementation validates the central directory size before handling the zip64 end of central directory record, and if the central directory size is `0xFFFFFFFF` the following check will fail:

https://github.com/classgraph/classgraph/blob/ad7ec172b59c3012e3296ec78d68f0436f2b8d0e/src/main/java/nonapi/io/github/classgraph/fastzipfilereader/LogicalZipFile.java#L489-L492

zip implementations often write the real central directory size in the non-zip64 end of central directory record as long as it fits, even if there is also a zip64 end of central directory record. But it is valid for an implementation to always write `0xFFFFFFFF` to this field when writing zip64 output.

---

The following repro creates an archive with a central directory size set to `0xFFFFFFFF`.

`java.util.ZipFile` and `java.util.ZipInputStream` can read the archive:

```
ZipFile
hello
ZipInputStream
hello
```

ClassGraph fails to scan the same jar:

```
2024-04-03T16:09:39.969-0700	ClassGraph	Opening classpath elements (took 0.007048 sec)
2024-04-03T16:09:39.972-0700	ClassGraph	-- Opening classpath element tmp.jar
2024-04-03T16:09:39.972-0700	ClassGraph	---- Opening jar: tmp.jar
2024-04-03T16:09:39.975-0700	ClassGraph	------ Could not open jarfile tmp.jar : java.io.IOException: Could not get logical zipfile tmp.jar : java.io.IOException: Could not get toplevel slice [tmp.jar ; byte range: 0..194 / 194]
```

<details>

<summary>Code for the repro</summary>

```java
import static java.nio.charset.StandardCharsets.UTF_8;

import io.github.classgraph.ClassGraph;
import io.github.classgraph.Resource;
import io.github.classgraph.ScanResult;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.zip.CRC32;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

public class Zip64 {

  public static void main(String[] args) throws IOException {
    Path path = Files.createTempFile("tmp", ".jar");
    writeZip64(path);

    System.err.println("ZipFile");
    try (ZipFile zipFile = new ZipFile(path.toFile())) {
      zipFile.stream().forEach(ze -> System.err.println(ze.getName()));
    }

    System.err.println("ZipInputStream");
    try (ZipInputStream zis = new ZipInputStream(Files.newInputStream(path))) {
      ZipEntry ze;
      while ((ze = zis.getNextEntry()) != null) {
        System.err.println(ze.getName());
      }
    }

    try (ScanResult scanResult =
        new ClassGraph()
            .overrideClassLoaders(new URLClassLoader(new URL[] {path.toUri().toURL()}))
            .acceptJars(path.getFileName().toString())
            .verbose()
            .scan()) {
      scanResult
          .getAllResources()
          .forEachByteArray(
              (Resource res, byte[] content) -> {
                System.err.println(res.getPath());
              });
    }
  }

  private static void writeEntry(ZipOutputStream zos, String name) throws IOException {
    ZipEntry ze = new ZipEntry(name);
    ze.setMethod(ZipEntry.STORED);
    ze.setSize(0);
    ze.setCrc(0);
    zos.putNextEntry(ze);
    zos.closeEntry();
  }

  static void writeZip64(Path path) throws IOException {

    ByteBuffer buf = ByteBuffer.allocate(1000);
    buf.order(ByteOrder.LITTLE_ENDIAN);

    // The jar has a single entry named 'hello', with the value 'world'
    byte[] name = "hello".getBytes(UTF_8);
    byte[] value = "world".getBytes(UTF_8);
    int crc = crc32(value);

    int localHeaderPosition = buf.position();

    // local file header signature     4 bytes  (0x04034b50)
    buf.putInt((int) ZipFile.LOCSIG);
    // version needed to extract       2 bytes
    buf.putShort((short) 0);
    // general purpose bit flag        2 bytes
    buf.putShort((short) 0);
    // compression method              2 bytes
    buf.putShort((short) 0);
    // last mod file time              2 bytes
    buf.putShort((short) 0);
    // last mod file date              2 bytes
    buf.putShort((short) 0);
    // crc-32                          4 bytes
    buf.putInt(crc);
    // compressed size                 4 bytes
    buf.putInt(value.length);
    // uncompressed size               4 bytes
    buf.putInt(value.length);
    // file name length                2 bytes
    buf.putShort((short) name.length);
    // extra field length              2 bytes
    buf.putShort((short) 0);
    // file name (variable size)
    buf.put(name);
    // extra field (variable size)
    // file data
    buf.put(value);

    int centralDirectoryPosition = buf.position();

    // central file header signature   4 bytes  (0x02014b50)
    buf.putInt((int) ZipFile.CENSIG);
    // version made by                 2 bytes
    buf.putShort((short) 0);
    // version needed to extract       2 bytes
    buf.putShort((short) 0);
    // general purpose bit flag        2 bytes
    buf.putShort((short) 0);
    // compression method              2 bytes
    buf.putShort((short) 0);
    // last mod file time              2 bytes
    buf.putShort((short) 0);
    // last mod file date              2 bytes
    buf.putShort((short) 0);
    // crc-32                          4 bytes
    buf.putInt(crc);
    // compressed size                 4 bytes
    buf.putInt(value.length);
    // uncompressed size               4 bytes
    buf.putInt(value.length);
    // file name length                2 bytes
    buf.putShort((short) name.length);
    // extra field length              2 bytes
    buf.putShort((short) 0);
    // file comment length             2 bytes
    buf.putShort((short) 0);
    // disk number start               2 bytes
    buf.putShort((short) 0);
    // internal file attributes        2 bytes
    buf.putShort((short) 0);
    // external file attributes        4 bytes
    buf.putInt(0);
    // relative offset of local header 4 bytes
    buf.putInt(localHeaderPosition);
    // file name (variable size)
    buf.put(name);

    int centralDirectorySize = buf.position() - centralDirectoryPosition;
    int zip64eocdPosition = buf.position();

    // zip64 end of central dir
    // signature                       4 bytes  (0x06064b50)
    buf.putInt(0x06064b50);
    // size of zip64 end of central
    // directory record                8 bytes
    buf.putLong(40 + 5);
    // version made by                 2 bytes
    buf.putShort((short) 0);
    // version needed to extract       2 bytes
    buf.putShort((short) 0);
    // number of this disk             4 bytes
    buf.putInt(0);
    // number of the disk with the
    // start of the central directory  4 bytes
    buf.putInt(0);
    // total number of entries in the
    // central directory on this disk  8 bytes
    buf.putLong(1);
    // total number of entries in the
    // central directory               8 bytes
    buf.putLong(1);
    // size of the central directory   8 bytes
    buf.putLong(centralDirectorySize);
    // offset of start of central
    // directory with respect to
    // offset of start of central
    // the starting disk number        8 bytes
    buf.putLong(centralDirectoryPosition);
    // zip64 extensible data sector    (variable size)
    buf.put((byte) 3);
    buf.putInt(42);

    // zip64 end of central dir locator
    // signature                       4 bytes  (0x07064b50)
    buf.putInt(0x07064b50);
    // number of the disk with the
    // start of the zip64 end of
    // central directory               4 bytes
    buf.putInt(0);
    // relative offset of the zip64
    // end of central directory record 8 bytes
    buf.putLong(zip64eocdPosition);
    // total number of disks           4 bytes
    buf.putInt(0);

    // end of central dir signature    4 bytes  (0x06054b50)
    buf.putInt((int) ZipFile.ENDSIG);
    // number of this disk             2 bytes
    buf.putShort((short) 0);
    // number of the disk with the
    // start of the central directory  2 bytes
    buf.putShort((short) 0);
    // total number of entries in the
    // central directory on this disk  2 bytes
    buf.putShort((short) 1);
    // total number of entries in
    // the central directory           2 bytes
    buf.putShort((short) 1);
    // size of the central directory   4 bytes
    buf.putInt(0xffffffff);
    // offset of start of central
    // directory with respect to
    // the starting disk number        4 bytes
    buf.putInt(0xffffffff);
    // .ZIP file comment length        2 bytes
    buf.putShort((short) 0);
    // .ZIP file comment       (variable size)

    byte[] bytes = new byte[buf.position()];
    buf.rewind();
    buf.get(bytes);
    Files.write(path, bytes);
  }

  static int crc32(byte[] value) {
    CRC32 crc32 = new CRC32();
    crc32.update(value);
    return (int) crc32.getValue();
  }
}
```

</details>

**Comment by lukehutch** [comment]
Liam, this is the sort of unbelievably high-quality bug report that I love to see! Thank you for tracking this down, providing a detailed analysis, and providing a fix.

(I see you are a Googler -- I am a Xoogler -- is ClassGraph in use at Google?)

I'll go ahead and merge your PR. Thank you!

**Comment by lukehutch** [comment]
Fixed and released in version 4.8.170. Much appreciated!

**Comment by cushon** [comment]
Thanks for the quick turnaround! And yep, we're still using ClassGraph.
